### PR TITLE
Added tzinfo to requestContext fields

### PR DESCRIPTION
**Pull request by winguru**
Bruno,

Here are some changes which seem to correct the missing timezone behavior of the smartSummarize function.

I added the tzinfo as a value to the requestContext dictionary in app.py.
I used this tzinfo in functions.py to add tzinfo to the smartSummarize function.
I also noticed that some datapoints (usually the last one in my experience) are missing timestamp data in the smartSummarize function, so I added a quick check to ensure they are present.

This bug is discussed here:
https://github.com/brutasse/graphite-api/issues/51

I've tried to ensure that this patch passed tox testing.  Travis-CI testing still seems to fail based on DeprecatedWarning messages in the setuptools module.  I have patches to the tox.ini file if you are interested.

Thanks for your time and consideration!

--Geoff Thornton

### CODE REVIEW

File: graphite_api/functions.py [hunk 1]
```diff
@@ -2711,14 +2711,18 @@ def smartSummarize(requestContext, seriesList, intervalString, func='sum'):
 
     # Adjust the start time to fit an entire day for intervals >= 1 day
     requestContext = requestContext.copy()
+    # ISSUE #51: Added tzinfo to requestContext in graphite_api/app.py:328
+    tzinfo = requestContext.get('tzinfo', None)
```
Comment (brutasse) [inline graphite_api/functions.py]: You can directly do `tzinfo = requestContext['tzinfo']`: it's guaranteed to be in the request context (and the comment above is unnecessary)

**Comment by brutasse** [comment]
Thanks Geoff! I made a minor comment. If you can address it (simply make a new commit and push, it'll update the pull request.

It would be great to have a test case that reproduces the issue.

File: graphite_api/functions.py (hunk 1 above)
Comment (winguru) [inline graphite_api/functions.py]: Bruno, I originally did the direct key request as you suggested, but the unit test failed for smartSummarize as it provided data directly to the function without any tzinfo key.  I figured I'd rather error on the side of caution and not change your unit tests as well ;-)

I can easily remove the comment, I just added it for notation...

Thanks!

--Geoff Thornton

**Comment by winguru** [comment]
I'll see if I can do a test case.  I was testing out this code with InfuxDB as the backend using a forked backend, but if you'd like me to work more on providing a specific test case / unit test, let me know and I'll work on it...

File: graphite_api/functions.py (hunk 1 above)
Comment (brutasse) [inline graphite_api/functions.py]: Ah right -- the unit test should reproduce what the app does, so if tzinfo is provided to all functions then tests should do the same… Feel free to update tests when this parameter is required :)

**Comment by brutasse** [comment]
Here's a minimal addition to `test_smart_summarize` that triggers the error and that should pass with your patch. Feel free to integrate it directly in your pull request:

```diff
diff --git a/tests/test_functions.py b/tests/test_functions.py
index 43ed644..f9efff8 100644
--- a/tests/test_functions.py
+++ b/tests/test_functions.py
@@ -1007,6 +1007,10 @@ class FunctionsTest(TestCase):
         summ = functions.smartSummarize(ctx, series, '5s', 'min')[0]
         self.assertEqual(summ[:3], [42, 47, 52])
 
+        # Higher time interval should not trigger timezone errors
+        functions.smartSummarize(ctx, series, '100s', 'min')[0]
+
     def test_summarize(self):
         series = self._generate_series_list(config=[list(range(99)) + [None]])
 

```

**Comment by winguru** [comment]
Bruno,

I've added your additional test into the test_smart_summarize function.  I had to import pytz locally (which I normally don't like doing) as I couldn't find any established tzinfo data in the testing modules.  If you have a better suggestion, I'm all ears ;-)

**Comment by winguru** [comment]
Bruno,

Let me know if there is anything else you'd like to see added or modified... I'm happy to do whatever you think is best regarding this changeset.

Thanks!

--Geoff Thornton

**Comment by brutasse** [comment]
Thanks Geoff, looks great! Why did you have to import `pytz` though? I already see it being imported at the top of the file.

**Comment by brutasse** [comment]
Ah wait, it's not present in your repository but in master it's here.

Could you update your fork to include the upstream changes?

```
git remote add upstream https://github.com/brutasse/graphite-api.git
git fetch upstream
git merge upstream/master
git push
```

**Comment by winguru** [comment]
Ah, thanks for the catch!  I thought I was updated to the latest master, but I'll update right now and ensure I'm in sync with your code ;-)

Thanks again!!

**Comment by winguru** [comment]
When I ran those commands you gave me, it showed that your upstream was already added.  When I tried to push, I got a message that it was already up to date:

```
> git remote add upstream https://github.com/brutasse/graphite-api.git
fatal: remote upstream already exists.

> git fetch upstream
> git merge upstream/master
Already up-to-date.

> git push
Everything up-to-date
```

However, I was able to run this command and merge your upstream master:
`git pull https://github.com/brutasse/graphite-api.git`

I then pushed the changes to origin/master and it looks like I am up-to-date now ;-)

The changeset seems to reflect the proper delta and I *do* see the import pytz at the top of the file.

Thanks for all of you help and guidance!!

**Comment by winguru** [comment]
Okay looks like all of the build tests passed ;-)

I'm still learning how to collaborate with others in GitHub, so thanks again for your patience... I think the problem was that Dieterbe/graphite-api is is 106 commits behind brutasse:master.  That probably got me out of sync.  Normally I do work with git and GitHub on my own projects, so I don't have a lot of experience with forking other's code and requesting pull requests.

I am very thankful for your continued involvement and I'm sure I speak on behalf of the community that we appreciate the time and effort you've spent in creating and maintaining this project ;-)

Thanks again!

--Geoff Thornton

**Comment by brutasse** [comment]
Great! I'm glad this project is useful. Thanks for tackling this issue!

Bruno

**Pull request merged**